source-git-monorepos/README.md

```diff
@@ -50,21 +89,369 @@ packages:
+
+### packit push-updates
+
+Should be updated similar to `packit create-update`.
+
+### packit status
+
+The command should display status for all packages in a monorepo.
+
+### packit validate-config
 
-For **backwards compatibility**: if the top-level `packages` key is missing,
+Check if the paths specified in `package_paths` do exist in the monorepo.
+
+Check if the package names (`downstream_package_name`) are correct and indeed
+exist in the distribution. See [packit/packit#88].
+
+### packit prepare-sources
+
+I'm not entirely sure my understanding of what this command does is correct 😅
+
+The command should be updated similar to `packit srpm` (?), so that it's possible
+to limit changes only to select packages.
+
+## Impact on Packit API
+
+### Internal Design
+
+With monorepos needs to be taught to map a single upstream repository
+(`PackitAPI.up`) to multiple dist-git/downstream repositories. This means that
+`PackitAPI.dg` should become a dictionary holding the objects for each
+dist-git repository.
+
+In order to make the transition easier, we could keep `.dg` as it is, and make
+it to return the dist-git repo for the first (and only) package. If there are
+more packages defined, an exception should be raised. This way parts of the
+API and CLI which were not yet updated to support monorepos, could continue to
+work, but would fail if used with a monorepo configuration specifying multiple
+packages.
+
+### Inheritance (Overriding) of Configuration Keys
+
+One can override configuration defined on the top-level of `.packit.yaml` in a
+package-object. That is, all top-level values are inherited by package-objects.
+
+Configuration values can be further overridden in job objects, but only for
+the packages the job references (if any.)
+
+```mermaid
+flowchart BT
+    C(Job) --> B(Package)
+    B --> A(Top-level)
+```
+
+A (not very meaningful) example:
+
+```yaml
+srpm_build_deps:
+  - python3-pip
+  - python3-setuptools
+packages:
+  subpackage1:
+    package_paths:
+      - package1
+    srpm_build_deps:
+      - python3-pip
+      - python3-setuptools
+      - python3-packaging
+  subpackage2:
+    package_paths:
+      - package2
+jobs:
+  - job: copr_build
+    trigger: pull_request
+    srpm_build_deps:
+      - python3-pip
+      - python3-setuptools
+      - cowsay
+    targets:
+      - fedora-all
+  - job: copr_build
+    trigger: release
+    packages:
+      - subpackage1
+    targets:
+      - fedora-all
+      - epel-9
+```
+
+In the example above when building PRs in Copr, the `srpm_build_deps` defined
+in the job would be used for both packages.
+
+On the other hand, when building releases in Copr, only `subpackage1` is going
+to be built, and the `srpm_build_deps` will be the ones defined for the
+package in the `packages` section.
+
+The example above could be made more useful if extending lists would be
+supported by Packit. (YAML doesn't support this, one gets nested lists when
```

Comment:
fmf… but that would be another story I guess

Reply:
Yes, it would another story. This could be even done separately from this whole monorepo-saga. Stackowerflow suggested something like:

```yaml
list1: &listref
- element1
- element2
list2: !flatten
- *listref
- additional-element
```

We would need to write the code to flatten `[['element1', 'element2'], 'additional-element']`, and set a trigger to do it when the `!flatten` tag is met.